File: front/src/services/agent.ts
import apiClient from './api';

export interface AgentModelSettings {
  temperature: number;
  top_p: number;
  frequency_penalty: number;
  presence_penalty: number;
  max_tokens: number;
}

export interface ToolConfig {
  enabled: boolean;
  name: string;
  api_key?: string;
  config?: Record<string, any>;
}

export interface Agent {
  id: number;
  user_id: number;
  system_prompt: string;
  model: string;
  max_memory: number;
  model_settings: AgentModelSettings;
  tools_enabled?: Record<string, ToolConfig>;
  created_at: string;
  updated_at: string;
}

export interface ApiResponse<T> {
  code: number;
  msg: string;
  data: T;
  errors: any;
  timestamp: string;
  request_id: string;
}

// ===================== 新的API端点 - 用户唯一Agent =====================

// 获取当前用户的AI助手
const getMyAgent = async (): Promise<Agent | null> => {
  try {
    console.log('发起getMyAgent请求');
    const response = await apiClient.get<ApiResponse<Agent>>('/agent/my-agent');
    console.log('getMyAgent请求成功');
    return response.data.data;
  } catch (error: any) {
    if (error.response?.status === 404) {
      console.log('用户还没有AI助手');
      return null;
    }
    console.error('获取用户AI助手失败:', error);
    return null;
  }
};

// 创建或更新当前用户的AI助手
const createOrUpdateMyAgent = async (agentData: Omit<Agent, 'id' | 'user_id' | 'created_at' | 'updated_at'>): Promise<Agent | null> => {
  try {
    console.log('发起createOrUpdateMyAgent请求');
    const response = await apiClient.post<ApiResponse<Agent>>('/agent/my-agent', agentData);
    console.log('createOrUpdateMyAgent请求成功');
    return response.data.data;
  } catch (error) {
    console.error('创建或更新AI助手失败:', error);
    return null;
  }
};

// 更新当前用户的AI助手
const updateMyAgent = async (agentData: Partial<Omit<Agent, 'id' | 'user_id' | 'created_at' | 'updated_at'>>): Promise<Agent | null> => {
  try {
    console.log('发起updateMyAgent请求');
    const response = await apiClient.put<ApiResponse<Agent>>('/agent/my-agent', agentData);
    console.log('updateMyAgent请求成功');
    return response.data.data;
  } catch (error) {
    console.error('更新AI助手失败:', error);
    return null;
  }
};

// 删除当前用户的AI助手
const deleteMyAgent = async (): Promise<boolean> => {
  try {
    console.log('发起deleteMyAgent请求');
    const response = await apiClient.delete<ApiResponse<{deleted_agent_id: number}>>('/agent/my-agent');
    console.log('deleteMyAgent请求成功');
    return true;
  } catch (error) {
    console.error('删除AI助手失败:', error);
    return false;
  }
};

// 获取公开的AI助手列表
const getPublicAgents = async (skip = 0, limit = 100): Promise<Agent[]> => {
  try {
    console.log('发起getPublicAgents请求');
    const response = await apiClient.get<ApiResponse<Agent[]>>('/agent/public-agents', {
      params: { skip, limit }
    });
    console.log(`getPublicAgents请求成功，数据项: ${response.data.data.length}`);
    return response.data.data;
  } catch (error) {
    console.error('获取公开AI助手列表失败:', error);
    return [];
  }
};

// ===================== 旧的API端点（保持向后兼容，但功能受限） =====================

// 获取用户能看到的所有Agent（自己的+所有公开的）
const getAllAgents = async (skip = 0, limit = 100): Promise<Agent[]> => {
  try {
    console.log(`发起getAllAgents请求`);
    // 不传入is_public，获取用户能看到的所有agents（包括自己的和所有公开的）
    const response = await apiClient.get<ApiResponse<Agent[]>>('/agent/agents', {
      params: { skip, limit }
    });
    console.log(`getAllAgents请求成功，数据项: ${response.data.data.length}`);
    return response.data.data;
  } catch (error) {
    console.error('获取所有Agent列表失败:', error);
    return [];
  }
};

// 获取用户自己的所有agent列表（现在只会返回一个agent）
const getUserAgents = async (skip = 0, limit = 100): Promise<Agent[]> => {
  try {
    const response = await apiClient.get<ApiResponse<Agent[]>>(`/agent/agents`, {
      params: { is_public: false, skip, limit }
    });
    return response.data.data;
  } catch (error) {
    console.error('获取用户Agent列表失败:', error);
    return [];
  }
};

// 获取Agent详情
const getAgentDetail = async (agentId: number): Promise<Agent | null> => {
  try {
    const response = await apiClient.get<ApiResponse<Agent>>(`/agent/agents/${agentId}`);
    return response.data.data;
  } catch (error) {
    console.error(`获取Agent(ID:${agentId})详情失败:`, error);
    return null;
  }
};

// 创建Agent（已弃用，重定向到新接口）
const createAgent = async (agentData: Omit<Agent, 'id' | 'user_id' | 'created_at' | 'updated_at'>): Promise<Agent | null> => {
  console.warn('createAgent已弃用，使用createOrUpdateMyAgent代替');
  return createOrUpdateMyAgent(agentData);
};

// 更新Agent（需要检查权限）
const updateAgent = async (agentId: number, agentData: Partial<Agent>): Promise<Agent | null> => {
  try {
    const response = await apiClient.put<ApiResponse<Agent>>(`/agent/agents/${agentId}`, agentData);
    return response.data.data;
  } catch (error) {
    console.error(`更新Agent(ID:${agentId})失败:`, error);
    return null;
  }
};

// 删除Agent（需要检查权限）
const deleteAgent = async (agentId: number): Promise<boolean> => {
  try {
    const response = await apiClient.delete<ApiResponse<{deleted_agent_id: number}>>(`/agent/agents/${agentId}`);
    return true;
  } catch (error) {
    console.error(`删除Agent(ID:${agentId})失败:`, error);
    return false;
  }
};

// 获取可用模型列表
const getAvailableModels = async (): Promise<string[]> => {
  try {
    const response = await apiClient.get<ApiResponse<string[]>>('/agent/models');
    return response.data.data;
  } catch (error) {
    console.error('获取可用模型列表失败:', error);
    return [];
  }
};

// 基于公开Agent创建自己的Agent（从模板复制设置）
const createAgentFromTemplate = async (templateAgentId: number): Promise<Agent | null> => {
  try {
    // 先获取模板Agent的详情
    const templateAgent = await getAgentDetail(templateAgentId);
    if (!templateAgent) {
      console.error(`模板Agent(ID:${templateAgentId})不存在或无法访问`);
      return null;
    }
    
    // 检查当前用户是否已有Agent
    const existingAgent = await getMyAgent();
    if (existingAgent) {
      // 如果已有Agent，询问是否覆盖
      const shouldOverwrite = confirm(`您已有AI助手，是否要用模板覆盖当前设置？`);
      if (!shouldOverwrite) {
        return null;
      }
    }
    
    // 创建新Agent，复制模板Agent的关键设置
    const newAgentData = {
      system_prompt: templateAgent.system_prompt,
      model: templateAgent.model,
      max_memory: templateAgent.max_memory,
      model_settings: templateAgent.model_settings,
      tools_enabled: templateAgent.tools_enabled
    };
    
    // 创建或更新用户的Agent
    const result = await createOrUpdateMyAgent(newAgentData);
    return result;
  } catch (error) {
    console.error(`基于模板创建Agent失败:`, error);
    return null;
  }
};

export default {
  // 新的推荐API
  getMyAgent,
  createOrUpdateMyAgent,
  updateMyAgent,
  deleteMyAgent,
  getPublicAgents,
  
  // 旧的API（保持兼容性）
  getAllAgents,
  getUserAgents,
  getAgentDetail,
  createAgent,
  updateAgent,
  deleteAgent,
  getAvailableModels,
  createAgentFromTemplate
}; 

File: backend/api/v1/endpoints/agent.py
from fastapi import APIRouter, Depends, HTTPException, Path, Body, Request, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Dict, Any, Optional
from datetime import datetime

from backend.db.session import get_async_session
from backend.api.deps import get_current_active_user, get_db
from backend.models.user import User
from backend.core.response import SuccessResponse
from backend.crud.agent import agent
from backend.schemas.agent import AgentCreate, AgentUpdate, AgentInDB, AgentListResponse
from backend.utils.logging import api_logger
from backend.core.config import settings
from backend.utils.id_converter import IDConverter

router = APIRouter()

def agent_to_dict(db_agent) -> Dict[str, Any]:
    """将Agent模型转换为字典"""
    return {
        "id": db_agent.public_id,
        "user_id": db_agent.user.public_id if db_agent.user else None,
        "system_prompt": db_agent.system_prompt,
        "model": db_agent.model,
        "max_memory": db_agent.max_memory,
        "model_settings": db_agent.model_settings,
        "tools_enabled": db_agent.tools_enabled,
        "created_at": db_agent.created_at.isoformat() if db_agent.created_at else None,
        "updated_at": db_agent.updated_at.isoformat() if db_agent.updated_at else None
    }


@router.get("/models")
async def list_models(
    request: Request,
    current_user: User = Depends(get_current_active_user),
):
    """
    获取可用的模型列表
    """
    api_logger.info(f"获取可用模型列表: {current_user.username}")
    
    # 返回一些常用模型作为参考，但不再限制用户使用的模型
    suggested_models = [
        "gpt-4o", 
        "claude-3-7-sonnet-20250219",
        'anthropic/claude-sonnet-4',
        'deepseek-r1-vol',
        # 添加默认模型
        settings.OPENAI_MODEL
    ]
    
    # 去除重复项
    suggested_models = list(set(suggested_models))
    
    return SuccessResponse(
        data=suggested_models,
        msg="获取模型列表成功",
        request_id=getattr(request.state, "request_id", None)
    )


@router.get("/my-agent", response_model=AgentInDB)
async def get_my_agent(
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    获取当前用户的AI助手
    """
    api_logger.info(f"获取用户AI助手: {current_user.username}")
    
    user_agent = await agent.get_user_agent(db, user_id=current_user.id)
    
    if not user_agent:
        api_logger.warning(f"用户没有AI助手: {current_user.username}")
        raise HTTPException(status_code=404, detail="您还没有AI助手，请先创建")
    
    # 转换为字典
    agent_dict = agent_to_dict(user_agent)
    
    return SuccessResponse(
        data=agent_dict,
        msg="获取AI助手成功",
        request_id=getattr(request.state, "request_id", None)
    )


@router.post("/my-agent", response_model=AgentInDB)
async def create_or_update_my_agent(
    request: Request,
    agent_in: AgentCreate = Body(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    创建或更新当前用户的AI助手
    """
    api_logger.info(f"创建或更新AI助手: 用户: {current_user.username}, 模型: {agent_in.model}")
    
    try:
        # 创建或更新用户的Agent
        db_agent = await agent.create_or_update_agent(db, user_id=current_user.id, obj_in=agent_in)
        
        # 转换为字典以确保JSON可序列化
        agent_dict = agent_to_dict(db_agent)
        
        return SuccessResponse(
            data=agent_dict,
            msg="AI助手保存成功",
            request_id=getattr(request.state, "request_id", None)
        )
    except ValueError as e:
        api_logger.error(f"创建或更新AI助手失败: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))


@router.put("/my-agent", response_model=AgentInDB)
async def update_my_agent(
    request: Request,
    agent_in: AgentUpdate = Body(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    更新当前用户的AI助手
    """
    api_logger.info(f"更新AI助手, 用户: {current_user.username}")
    
    # 获取用户的Agent
    user_agent = await agent.get_user_agent(db, user_id=current_user.id)
    if not user_agent:
        api_logger.warning(f"用户没有AI助手: {current_user.username}")
        raise HTTPException(status_code=404, detail="您还没有AI助手，请先创建")
    
    # 更新Agent
    updated_agent = await agent.update_agent(db, agent_id=user_agent.public_id, obj_in=agent_in)
    
    # 转换为字典
    agent_dict = agent_to_dict(updated_agent)
    
    return SuccessResponse(
        data=agent_dict,
        msg="AI助手更新成功",
        request_id=getattr(request.state, "request_id", None)
    )


@router.delete("/my-agent")
async def delete_my_agent(
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    删除当前用户的AI助手
    """
    api_logger.info(f"删除AI助手, 用户: {current_user.username}")
    
    # 获取用户的Agent
    user_agent = await agent.get_user_agent(db, user_id=current_user.id)
    if not user_agent:
        api_logger.warning(f"用户没有AI助手: {current_user.username}")
        raise HTTPException(status_code=404, detail="您还没有AI助手")
    
    # 软删除Agent
    await agent.remove(db, id=user_agent.id)
    
    return SuccessResponse(
        data={"deleted_agent_id": user_agent.public_id},
        msg="AI助手删除成功",
        request_id=getattr(request.state, "request_id", None)
    )


@router.get("/public-agents", response_model=List[AgentListResponse])
async def list_public_agents(
    request: Request,
    skip: int = Query(0, description="分页起始位置"),
    limit: int = Query(100, description="分页大小，最大100条"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    获取公开的AI助手列表
    """
    api_logger.info(f"获取公开AI助手列表, 用户: {current_user.username}")
    
    public_agents = await agent.get_public_agents(db, skip=skip, limit=limit)
    
    # 转换为字典列表
    agent_list = [agent_to_dict(a) for a in public_agents]
    
    return SuccessResponse(
        data=agent_list,
        msg="获取公开AI助手列表成功",
        request_id=getattr(request.state, "request_id", None)
    )


@router.get("/agents/{agent_id}", response_model=AgentInDB)
async def get_agent_detail(
    request: Request,
    agent_id: str = Path(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    获取指定Agent的详细信息
    """
    api_logger.info(f"获取Agent详情: 用户: {current_user.username}, Agent ID: {agent_id}")
    
    # 直接使用public_id获取Agent（CRUD层已支持）
    user_agent = await agent.get_agent_by_id(db, agent_id=agent_id)
    
    if not user_agent:
        api_logger.warning(f"Agent不存在: {agent_id}")
        raise HTTPException(status_code=404, detail="Agent不存在")
    
    # 检查Agent是否属于当前用户或是公开的
    if user_agent.user_id != current_user.id:
        api_logger.warning(f"用户无权访问Agent: 用户: {current_user.username}, Agent ID: {agent_id}")
        raise HTTPException(status_code=403, detail="无权访问此Agent")
    
    # 转换为字典
    agent_dict = agent_to_dict(user_agent)
    
    return SuccessResponse(
        data=agent_dict,
        msg="获取Agent详情成功",
        request_id=getattr(request.state, "request_id", None)
    )


@router.put("/agents/{agent_id}", response_model=AgentInDB)
async def update_agent_detail(
    request: Request,
    agent_id: str = Path(...),
    agent_in: AgentUpdate = Body(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    更新指定Agent的信息
    """
    api_logger.info(f"更新Agent: 用户: {current_user.username}, Agent ID: {agent_id}")
    
    try:
        # 检查Agent是否属于当前用户（CRUD层已支持public_id）
        ownership_check = await agent.check_agent_ownership(db, agent_id=agent_id, user_id=current_user.id)
        if not ownership_check:
            api_logger.warning(f"用户无权更新Agent: 用户: {current_user.username}, Agent ID: {agent_id}")
            raise HTTPException(status_code=403, detail="无权更新此Agent")
        
        # 更新Agent（CRUD层已支持public_id）
        updated_agent = await agent.update_agent(db, agent_id=agent_id, obj_in=agent_in)
        
        if not updated_agent:
            raise HTTPException(status_code=404, detail="Agent不存在")
        
        # 转换为字典
        agent_dict = agent_to_dict(updated_agent)
        
        return SuccessResponse(
            data=agent_dict,
            msg="Agent更新成功",
            request_id=getattr(request.state, "request_id", None)
        )
    except ValueError as e:
        api_logger.error(f"更新Agent失败: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/agents/{agent_id}")
async def delete_agent_detail(
    request: Request,
    agent_id: str = Path(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    删除指定Agent
    """
    api_logger.info(f"删除Agent: 用户: {current_user.username}, Agent ID: {agent_id}")
    
    # 检查Agent是否属于当前用户（CRUD层已支持public_id）
    ownership_check = await agent.check_agent_ownership(db, agent_id=agent_id, user_id=current_user.id)
    if not ownership_check:
        api_logger.warning(f"用户无权删除Agent: 用户: {current_user.username}, Agent ID: {agent_id}")
        raise HTTPException(status_code=403, detail="无权删除此Agent")
    
    # 转换为数据库内部ID进行删除（删除操作仍需要内部ID）
    db_agent_id = await IDConverter.get_agent_db_id(db, agent_id)
    if not db_agent_id:
        raise HTTPException(status_code=404, detail="Agent不存在")
    
    # 删除Agent
    success = await agent.remove(db, id=db_agent_id)
    if not success:
        raise HTTPException(status_code=404, detail="Agent不存在")
    
    return SuccessResponse(
        data={"deleted_agent_id": agent_id},  # 返回public_id
        msg="Agent删除成功", 
        request_id=getattr(request.state, "request_id", None)
    )


# 保留旧的API端点以向后兼容，但限制功能

@router.post("/agents", response_model=AgentInDB)
async def create_agent(
    request: Request,
    agent_in: AgentCreate = Body(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    创建新的Agent（已弃用，请使用 POST /my-agent）
    """
    api_logger.warning(f"使用已弃用的创建Agent接口: {current_user.username}")
    # 重定向到新接口
    return await create_or_update_my_agent(request, agent_in, db, current_user)


@router.get("/agents", response_model=List[AgentListResponse])
async def list_agents(
    request: Request,
    skip: int = Query(0, description="分页起始位置"),
    limit: int = Query(100, description="分页大小，最大100条"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    获取Agent列表（只返回用户自己的agent）
    """
    api_logger.info(f"获取用户的Agent列表: {current_user.username}")
    
    # 获取用户自己的Agent
    user_agent = await agent.get_user_agent(db, user_id=current_user.id)
    agents_list = [user_agent] if user_agent else []
    
    # 转换为字典列表
    agent_list = [agent_to_dict(a) for a in agents_list]
    
    return SuccessResponse(
        data=agent_list,
        msg="获取Agent列表成功",
        request_id=getattr(request.state, "request_id", None)
    ) 